docs/agent-development.md

```diff
@@ -384,3 +387,54 @@ Here's a list of the config options across all agents, their types, default valu
+
+1. Using the `GET` method, pass them as query parameters: `http://localhost:8200/config/v1/agents?service.name=opbeans&service.environment=production`
+2. Using the `POST` method, encode the parameters as a JSON object in the body, e.g. `{"service": {"name": "opbeans", "environment": "production"}}`
+
+The server will respond with a JSON object, where each key maps a config attribute to a string value. The string value should be interpreted the same as if it were passed in via an environment variable. Upon receiving these config changes, the agent will update its configuration dynamically, overriding any config previously specified. That is, config via Kibana takes highest precedence.
+
+To minimise the amount of work required by users, agents should aim to enable this feature by default. This excludes RUM, where there is a performance penalty.
+
+#### Caching
+
+As mentioned above, the server will cache config for each unique `service.name`, `service.environment` pair. The server will respond to config requests with two related response headers: [Etag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control).
+
+Agents should keep a record of the `Etag` value returned by the most successful config request, and then present it to future requests via the [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header. If the config has not changed, the server will respond with 304 (Not Modified).
+
+The `Cache-Control` header should contain a `max-age` directive, specifying the amount of time (in seconds) the response should be considered "fresh". Agents should use this to decide how long to wait before requesting config again. The server will respond with a `Cache-Control` header even if the request fails.
+
+#### Dealing with errors
+
+Agents must deal with various error scenarios, including:
+
+ - 7.3 servers where the Kibana connection is not enabled (server responds with 403)
+ - 7.3 servers where the Kibana connection is enabled, but unavailable (server responds with 503)
+ - 7.3 servers where the Kibana connection is enabled, available, but there is no matching agent config (server responds with 404)
+ - pre-7.3 servers that don't support the config endpoint (server responds with 404)
+ - any other error (server responds with 5xx)
+
+If the server responds with any 5xx, agents should log at error level. If the server responds with 4xx, agents are not required to log the response. Either central config is not available, not enabled, or there is no agent config to update. In all of these cases, there is no expectation that the agent should do anything, so logging is not useful.
```

Comment:
agents may log on debug log level

Reply:
changed "so logging is not useful" to "so logging is not necessary", and added a remark about agents being free to log at debug level